File: tooling/harness-eval/runner.ts
import type { Controls, Report, Run } from "./report-schema.ts";
import {
  type LoadedCase,
  fingerprint,
  loadCases,
  readSource,
} from "./sources.ts";
import {
  type PreparedFixture,
  collectObservations,
  prepareFixture,
} from "./fixture.ts";
import { readdirSync, rmSync } from "node:fs";
import { evaluate } from "./oracle.ts";
import { join } from "node:path";
import { validateReport } from "./evidence.ts";

const MAX_RUNS = 10;
type Execution = Omit<Run, "status" | "observations">;
type Executor = (
  fixture: PreparedFixture,
  prompt: string,
) => Promise<Execution> | Execution;
type SeriesOptions = Readonly<{
  agent: string;
  agentVersion: string;
  model: string;
  only: readonly string[];
  runs: number;
  controls: Controls;
  variant?: string;
}>;
type SeriesContext = Readonly<{
  repository: string;
  options: SeriesOptions;
  revision: string;
}>;
type EvaluatedCase = Readonly<{
  harness: Report["harness"];
  result: Report["cases"][number];
}>;

function runnerRevision(repository: string): string {
  const files = readdirSync(join(repository, "tooling/harness-eval"))
    .filter(
      (path) =>
        path.endsWith(".ts") &&
        !path.endsWith(".test.ts") &&
        path !== "test-support.ts",
    )
    .toSorted();
  return fingerprint(
    files
      .map(
        (path) =>
          `${path}\0${readSource(repository, `tooling/harness-eval/${path}`)}`,
      )
      .join("\0"),
  );
}

function gitRevision(repository: string): string {
  const result = Bun.spawnSync(["git", "-C", repository, "rev-parse", "HEAD"], {
    env: {
      PATH: "/usr/bin:/bin",
      GIT_CONFIG_NOSYSTEM: "1",
      GIT_CONFIG_GLOBAL: "/dev/null",
    },
  });
  if (result.exitCode !== 0) {
    throw new Error("Cannot identify harness Git revision");
  }
  return result.stdout.toString().trim();
}

async function observe(
  fixture: PreparedFixture,
  entry: LoadedCase,
  execute: Executor,
): Promise<Run> {
  const result = await execute(fixture, entry.prompt);
  try {
    const observations = collectObservations(fixture.observations);
    return {
      ...result,
      observations,
      status:
        result.error === null
          ? evaluate(entry.definition.oracle, observations)
          : "INVALID",
    };
  } catch {
    return {
      ...result,
      observations: [],
      status: "INVALID",
      error: "observation-invalid",
    };
  }
}

async function evaluateCase(
  context: SeriesContext,
  entry: LoadedCase,
  execute: Executor,
): Promise<EvaluatedCase> {
  const runs: Run[] = [];
  let identity: Report["harness"] | undefined = undefined;
  let fixtureRevision = "";
  for (let index = 0; index < context.options.runs; index += 1) {
    const fixture = prepareFixture(
      context.repository,
      entry,
      context.options.variant,
    );
    try {
      const harness = {
        gitRevision: context.revision,
        instructionFingerprint: fixture.instructionFingerprint,
        skillFingerprint: fixture.skillFingerprint,
        variant: context.options.variant ?? "Context Management",
      };
      if (
        identity !== undefined &&
        JSON.stringify(identity) !== JSON.stringify(harness)
      ) {
        throw new Error("Harness changed during evaluation");
      }
      identity = harness;
      ({ fixtureRevision } = fixture);
      runs.push(await observe(fixture, entry, execute));
    } finally {
      rmSync(fixture.root, { recursive: true, force: true });
    }
  }
  if (identity === undefined) {
    throw new Error("No evaluation runs");
  }
  return {
    harness: identity,
    result: {
      definition: entry.definition,
      sources: entry.sources,
      prompt: entry.prompt,
      promptFingerprint: entry.promptFingerprint,
      fixtureRevision,
      runs,
    },
  };
}

function selectCases(
  repository: string,
  options: SeriesOptions,
): readonly LoadedCase[] {
  const available = loadCases(repository);
  if (
    options.only.length === 0 ||
    new Set(options.only).size !== options.only.length ||
    options.only.some(
      (id) => !available.some((entry) => entry.definition.id === id),
    )
  ) {
    throw new Error("Explicit, unique known case selection required");
  }
  if (
    !Number.isInteger(options.runs) ||
    options.runs < 1 ||
    options.runs > MAX_RUNS
  ) {
    throw new Error("Runs must be between 1 and 10");
  }
  return available.filter((entry) =>
    options.only.includes(entry.definition.id),
  );
}

const limitations = [
  "Only Context Management and code-search are installed; this is not the full deployed harness.",
  "PATH shims observe supported commands, not internal skill loading or uninstrumented reads; bypasses can cause false negatives.",
  "Synthetic shims are not ColGrep quality or security tests; observations are not tamper-proof.",
  "Git revision plus content fingerprints identify the tested bytes, including uncommitted changes.",
  "Model is the requested ID; aliases may resolve differently. No statistical or causal uplift claim.",
  "No token ceiling is available; the wall-clock timeout bounds each run. Raw transcripts are discarded.",
] as const;

async function runSeries(
  repository: string,
  options: SeriesOptions,
  execute: Executor,
): Promise<Report> {
  const selected = selectCases(repository, options);
  const evaluated: EvaluatedCase[] = [];
  const revision = gitRevision(repository);
  const runner = runnerRevision(repository);
  for (const entry of selected) {
    evaluated.push(
      await evaluateCase({ repository, options, revision }, entry, execute),
    );
  }
  const [first] = evaluated;
  if (
    first === undefined ||
    evaluated.some(
      (entry) =>
        JSON.stringify(entry.harness) !== JSON.stringify(first.harness),
    ) ||
    runnerRevision(repository) !== runner
  ) {
    throw new Error("Evaluation sources changed during execution");
  }
  return validateReport({
    schemaVersion: 1,
    agent: options.agent,
    agentVersion: options.agentVersion,
    model: options.model,
    date: new Date().toISOString(),
    harness: first.harness,
    runnerRevision: runner,
    environment: {
      platform: process.platform,
      architecture: process.arch,
      bun: Bun.version,
    },
    controls: options.controls,
    runCount: options.runs,
    cases: evaluated.map((entry) => entry.result),
    limitations,
  });
}

export { runnerRevision, runSeries };
export type { Execution, Executor, SeriesOptions };

File: tooling/arnes/src/eval/runner.rs
use super::evidence::validate_report;
use super::fixture::Fixture;
use super::identity::{date, git_revision, runner_revision};
use super::live::{Codex, Execution, LiveOptions};
use super::observe::observe;
use super::report::{Controls, Environment, Harness, Report, ReportCase};
use super::sources::{LoadedCase, fingerprint, load_cases, read_source, section};
use std::collections::HashSet;
use std::path::Path;

pub struct SeriesOptions {
    pub model: String,
    pub only: Vec<String>,
    pub runs: u64,
    pub timeout_seconds: u64,
    pub reasoning_effort: String,
    pub variant: Option<String>,
}

pub fn run_live(repository: &Path, options: &SeriesOptions) -> Result<Report, String> {
    let selected = select_cases(repository, options)?;
    let codex = Codex::discover()?;
    let live = LiveOptions {
        model: options.model.clone(),
        reasoning_effort: options.reasoning_effort.clone(),
        timeout_seconds: options.timeout_seconds,
    };
    run_series(
        repository,
        options,
        selected,
        "codex",
        &codex.version,
        |fixture, entry| codex.execute(fixture, &entry.prompt, &live),
    )
}

pub fn run_smoke(repository: &Path) -> Result<Report, String> {
    let selected = load_cases(repository)?;
    let options = SeriesOptions {
        model: "none".to_owned(),
        only: selected
            .iter()
            .map(|entry| entry.definition.id.clone())
            .collect(),
        runs: 1,
        timeout_seconds: 120,
        reasoning_effort: "low".to_owned(),
        variant: None,
    };
    run_series(
        repository,
        &options,
        selected,
        "fixture-smoke",
        "1",
        super::smoke::execute,
    )
}

fn select_cases(repository: &Path, options: &SeriesOptions) -> Result<Vec<LoadedCase>, String> {
    let available = load_cases(repository)?;
    let unique: HashSet<_> = options.only.iter().collect();
    if unique.is_empty()
        || unique.len() != options.only.len()
        || unique
            .iter()
            .any(|id| !available.iter().any(|entry| &entry.definition.id == *id))
    {
        return Err("Explicit, unique known case selection required".to_owned());
    }
    if !(1..=10).contains(&options.runs) || !(1..=600).contains(&options.timeout_seconds) {
        return Err("Runs must be 1–10 and timeout seconds 1–600".to_owned());
    }
    Ok(available
        .into_iter()
        .filter(|entry| unique.contains(&entry.definition.id))
        .collect())
}

fn run_series(
    repository: &Path,
    options: &SeriesOptions,
    selected: Vec<LoadedCase>,
    agent: &str,
    version: &str,
    mut execute: impl FnMut(&Fixture, &LoadedCase) -> Result<Execution, String>,
) -> Result<Report, String> {
    let executable = std::env::current_exe().map_err(|error| error.to_string())?;
    let revision = runner_revision(&executable)?;
    let git = git_revision(repository)?;
    let mut identity = None;
    let mut cases = Vec::new();
    for entry in selected {
        let mut runs = Vec::new();
        for _ in 0..options.runs {
            let (fixture, harness) = prepare(repository, options, &entry, &executable, &git)?;
            if identity
                .as_ref()
                .is_some_and(|previous| previous != &harness)
            {
                return Err("Harness changed during evaluation".to_owned());
            }
            identity = Some(harness);
            runs.push(observe(&fixture, &entry, execute(&fixture, &entry)?));
        }
        let fixture_revision = fingerprint(format!(
            "{}\0{revision}",
            serde_json::to_string(&entry.fixture).map_err(|error| error.to_string())?
        ));
        cases.push(ReportCase {
            definition: entry.definition,
            prompt: entry.prompt,
            prompt_fingerprint: entry.prompt_fingerprint,
            sources: entry.sources,
            fixture_revision,
            runs,
        });
    }
    if runner_revision(&executable)? != revision {
        return Err("Evaluation executable changed during execution".to_owned());
    }
    let report = report(
        options,
        agent,
        version,
        identity.ok_or("No evaluation runs")?,
        revision,
        cases,
    )?;
    validate_report(&report)?;
    Ok(report)
}

fn prepare(
    repository: &Path,
    options: &SeriesOptions,
    entry: &LoadedCase,
    executable: &Path,
    git: &str,
) -> Result<(Fixture, Harness), String> {
    let instructions = match &options.variant {
        Some(path) => read_source(repository, path)?,
        None => section(
            &read_source(repository, "harness/AGENTS.md")?,
            "Context Management",
        )?
        .to_owned(),
    };
    let skill = read_source(repository, "harness/skills/code-search/SKILL.md")?;
    let harness = Harness {
        git_revision: git.to_owned(),
        instruction_fingerprint: fingerprint(&instructions),
        skill_fingerprint: fingerprint(&skill),
        variant: options
            .variant
            .clone()
            .unwrap_or_else(|| "Context Management".to_owned()),
    };
    Ok((
        Fixture::prepare(&entry.fixture.files, &instructions, &skill, executable)?,
        harness,
    ))
}

fn report(
    options: &SeriesOptions,
    agent: &str,
    version: &str,
    harness: Harness,
    revision: String,
    cases: Vec<ReportCase>,
) -> Result<Report, String> {
    Ok(Report {
        schema_version: 1, agent: agent.to_owned(), agent_version: version.to_owned(),
        model: options.model.clone(), date: date()?, harness, runner_revision: revision,
        environment: Environment::Rust {
            platform: std::env::consts::OS.to_owned(), architecture: std::env::consts::ARCH.to_owned(),
            runtime: format!("arnes-rust-{}", env!("CARGO_PKG_VERSION")),
        },
        controls: Controls {
            sandbox: "workspace-write".to_owned(), network: false,
            tools: "shell-with-synthetic-cat-rg-fd-colgrep-v1".to_owned(),
            timeout_seconds: options.timeout_seconds, reasoning_effort: options.reasoning_effort.clone(), token_budget: (),
        },
        run_count: options.runs as usize, cases,
        limitations: [
            "Only Context Management and code-search are installed; this is not the full deployed harness.",
            "PATH shims observe supported commands, not internal skill loading or uninstrumented reads; bypasses can cause false negatives.",
            "Synthetic shims are not ColGrep quality or security tests; observations are not tamper-proof.",
            "Git revision plus content fingerprints identify the tested bytes, including uncommitted changes.",
            "Model is the requested ID; aliases may resolve differently. No statistical or causal uplift claim.",
            "No token ceiling is available; the wall-clock timeout bounds each run. Raw transcripts are discarded.",
        ].map(str::to_owned).to_vec(),
    })
}
